Add basic ability to configure MachineSets (#1)
appuio/component-openshift4-nodes: docs/modules/ROOT/pages/references/parameters.adoc

= Parameters

The parent key for all of the following parameters is `openshift4_nodes`.

[CAUTION]
====
This component relies on deep merge of values from several parameters and hierarchy layers.

> This works pretty strait forward for scalar values and dictionaries.

Reviewer: ```suggestion
This works pretty straightforward for scalar values and dictionaries.
```

Values of arrays will be appended to each other.

> There is no way to override values of set on lover precedence location.

Reviewer: ```suggestion
There is no way to override values which were set on a lower precedence location.
```

====

> == `availabilityZones`

Reviewer: I'd propose to move this to `parameters.cloud.availabilityZones`, so it's together with `cloud.provider` and `cloud.region`.

Author: As discussed: we will keep this parameter in the component scope. Assignments from a global scope can be made within the hierarchy. This reduces the amount of implicit dependencies and has some more advantages (e.g. facts have highest precedence but when having them assigned to a component parameter, they can be overwritten again if need be).

The first item of this list will also be used to place a MachineSet without an explicit zone defined.

It's suggested to define its values higher up the config hierarchy.
Most probably the one of the cloud region.

== `defaultSpecs`

[horizontal]
type:: dictionary
default:: Sensible defaults for a growing number of cloud providers.

A dictionary holding the default values applied to each `machinesets.machine.openshift.io` object created by this component.

The top level keys are the names of cloud providers as reported by the `cloud.provider` fact.
The values can be everything that's accepted in the `spec` field of a `machinesets.machine.openshift.io` object.

== `infrastructureID`

[horizontal]
type:: string
default:: undefined

This is the 12 character infrastructure ID given to a cluster by the OpenShift 4 installer.
Use the following command to retrieve this from the cluster:

[code,bash]
----
oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
----

See also https://github.com/appuio/component-openshift4-nodes/issues/2[Get a clusters infrastructure ID as a fact].

[TIP]
====

> This is most like to be configured on cluster level itself.

Reviewer: ```suggestion
This is most likely to be configured on cluster level itself.
```

> Configuring that higher up in the hierarchy, can result in unexpected behavior.

Reviewer: ```suggestion
Configuring this higher up in the hierarchy can result in unexpected behavior.
```

Each set of machines has the values described below.

=== `instanceType`

=== `multiAz`

[horizontal]
type:: boolean
default:: false

A machine set will be placed in one single availability zone.
If not specified otherwise, the first entry of `availabilityZones` will be used.
When set to true, a MachineSet will be created for each zone listed in `availabilityZones`.
The replicas of the generated MachineSets will be calculated.
The `replicas` given will be divided by the cont of zones in `availabilityZones` rounded up.

See also https://github.com/appuio/component-openshift4-nodes/issues/3[Effective replica count of multi zone machines can be higher than the requested one]

=== `replicas`

[horizontal]
type:: number
default:: 1

The number of machines to create.
When `multiAZ` is set to `true`, the number given here will be divided so that each of the created MachineSets will get a fraction of replicas but the total of created machines will match the one requested here.

See also <<multiAZ>>.

[NOTE]
====
This value can also be set in <<spec>>.
If done so, the value in <<spec>> will win.
====

=== `role`

[horizontal]
type:: string
default:: worker

The role of the created Nodes.

> The value will be added as the `node-role.kubernetes.io/<role>: ""` label to node.

Reviewer: ```suggestion
The value will be added as the `node-role.kubernetes.io/<role>: ""` label to nodes.
```
